ambry-clustermap/src/main/java/com/github/ambry/clustermap/HelixBootstrapUpgradeUtil.java

```diff
@@ -788,6 +833,286 @@ private void migrateToPropertyStore() throws InterruptedException {
+          for (String resource : resources) {
+            // a. Get list of instances sharing partitions in this resource
+            IdealState idealState = resourceToIdealState.get(resource);
+            Set<String> partitions = idealState.getPartitionSet();
+            Set<String> instances = new HashSet<>();
+            for (String partition : partitions) {
+              instances.addAll(idealState.getInstanceSet(partition));
+            }
+            // b. Update instance config for each instance
+            for (String instance : instances) {
+              setupInstanceConfig(dcName, instance, resource, configAccessor, propertyStoreAdapter);
+            }
+            allInstances.addAll(instances);
+          }
+
+          // 3. No resource config to update for now
+
+          // 4. Validate configs
+          Map<String, Boolean> instanceValidationResultMap =
+              helixAdmin.validateInstancesForWagedRebalance(clusterName, new ArrayList<>(allInstances));
+          if (instanceValidationResultMap.containsValue(Boolean.FALSE)) {
+            throw new IllegalStateException("Instances are not configured correctly for waged rebalancer");
+          }
+          Map<String, Boolean> resourceValidationResultMap =
+              helixAdmin.validateResourcesForWagedRebalance(clusterName, new ArrayList<>(resources));
+          if (resourceValidationResultMap.containsValue(Boolean.FALSE)) {
+            throw new IllegalStateException("Resources are not configured correctly for waged rebalancer");
+          }
+
+          if (!dryRun) {
+            // Before
+            Map<String, Set<String>> replicaPlacementBefore = getReplicaPlacement(helixAdmin, resources);
+
+            // 5. Enter maintenance mode
+            helixAdmin.manuallyEnableMaintenanceMode(clusterName, true, "Migrating to Full auto",
+                Collections.emptyMap());
+
+            // 6. Update ideal state and enable waged rebalancer
+            for (String resource : resources) {
+              IdealState idealState = resourceToIdealState.get(resource);
+              idealState.setRebalanceMode(IdealState.RebalanceMode.FULL_AUTO);
+              // Set resource tag
+              idealState.setInstanceGroupTag(getResourceTag(resource));
+              // Set number of replicas
+              idealState.setReplicas(NUM_REPLICAS);
+              idealState.setMinActiveReplicas(MIN_ACTIVE_REPLICAS);
+              helixAdmin.updateIdealState(clusterName, resource, idealState);
+            }
+            helixAdmin.enableWagedRebalance(clusterName, new ArrayList<>(resources));
+
+            // 7. Exit maintenance mode
+            helixAdmin.manuallyEnableMaintenanceMode(clusterName, false, "Complete migrating to Full auto",
+                Collections.emptyMap());
+
+            // After
+            Map<String, Set<String>> replicaPlacementAfter = getReplicaPlacement(helixAdmin, resources);
```

Comment:
After exiting maintenance mode, it will take some time for helix to reshuffle the cluster, we have to wait for some time. 

There are two different ways to do this, one is to save the previous replica placement and create another command to just print out migration stats later. Another way is to wait for 30 seconds and check if the new replica placement is the same as the previous one and wait for another 30 seconds, util they are the same.

Reply:
Sure, I removed the logs for now. Like you said, I will try to add another command to print this.